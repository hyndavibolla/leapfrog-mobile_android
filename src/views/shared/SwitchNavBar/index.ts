export { default as SwitchNavBar, SwitchNavBarAlt } from './SwitchNavBar';
