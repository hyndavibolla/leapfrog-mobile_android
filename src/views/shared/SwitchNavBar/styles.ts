import { css } from '@emotion/native';

import { COLOR, FONT_SIZE } from '../../../constants';

export const styles = {
  container: css`
    align-items: center;
    justify-items: center;
  `,
  invalidContainer: css`
    border: 1px ${COLOR.RED} solid;
    border-radius: 4px;
  `,
  containerSwitch: css`
    flex-direction: row;
    align-items: center;
    justify-items: center;
  `,
  containerOptions: css`
    background-color: ${COLOR.MEDIUM_GRAY};
    flex: 1;
    border-radius: 50px;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
  `,
  toggleContainer: css`
    flex: 1;
    padding: 12px 28px;
    flex-direction: row;
    justify-content: center;
    align-items: center;
  `,
  toggle: css`
    border-radius: 21px;
  `,
  toggleActive: css`
    background-color: ${COLOR.WHITE};
    position: absolute;
    height: 100%;
  `,
  toggleInactive: css``,
  text: css`
    text-transform: uppercase;
    color: ${COLOR.BLACK};
    font-size: ${FONT_SIZE.TINY};
    letter-spacing: 1px;
    line-height: 14px;
  `,
  containerToggle: css`
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
  `,
  hiddenPill: css`
    width: 0;
    height: 0;
  `,
  altContainer: css`
    background-color: ${COLOR.WHITE};
  `,
  altActivePill: css`
    background-color: ${COLOR.PRIMARY_BLUE};
    border-radius: 8px;
  `,
  altActiveText: css`
    color: ${COLOR.WHITE};
  `,
  altInactiveText: css`
    color: ${COLOR.DARK_GRAY};
  `
};
