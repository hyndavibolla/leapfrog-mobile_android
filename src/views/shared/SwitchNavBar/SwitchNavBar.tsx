import React, { memo } from 'react';
import { View, TouchableHighlight, ViewStyle, TextStyle } from 'react-native';
import { View as AnimatedView } from 'react-native-animatable';

import { Text } from '../Text';
import { styles } from './styles';
import { FONT_FAMILY } from '../../../constants';
import { useTestingHelper } from '../../../utils/useTestingHelper';

export interface IOption {
  key: string;
  label: string;
}

export interface Props {
  optionList: IOption[];
  activeKey: string;
  onPress: (key: string) => void;
  containerStyle?: ViewStyle;
  activePillStyle?: ViewStyle;
  activeTextStyle?: TextStyle;
  inactiveTextStyle?: TextStyle;
  isInvalid?: boolean;
}

export const SwitchNavBar = ({ optionList, activeKey, onPress, containerStyle, activePillStyle, activeTextStyle, inactiveTextStyle, isInvalid }: Props) => {
  const { getTestIdProps } = useTestingHelper('switch-nav-bar');
  const optionWidth = 100 / optionList.length;
  const activeLeft = Math.max(optionList.findIndex(o => o.key === activeKey) * optionWidth, 0);
  const activePaddingPercentage = 2;

  return (
    <View style={[styles.container, isInvalid && styles.invalidContainer]}>
      <View style={styles.containerSwitch}>
        <View style={[styles.containerOptions, containerStyle, { padding: `${activePaddingPercentage / 2}%` }]}>
          <AnimatedView
            transition="left"
            duration={100}
            style={[
              styles.toggle,
              styles.toggleActive,
              activePillStyle,
              { width: `${optionWidth - activePaddingPercentage}%`, left: `${activeLeft + activePaddingPercentage}%` },
              [null, undefined].includes(activeKey) && styles.hiddenPill
            ]}
          />
          {optionList.map(({ key, label }) => (
            <TouchableHighlight
              {...getTestIdProps(`btn-${key}`)}
              key={key}
              underlayColor="transparent"
              style={styles.toggleContainer}
              onPress={() => onPress(key)}
            >
              <Text
                font={FONT_FAMILY.HEAVY}
                numberOfLines={1}
                ellipsizeMode="tail"
                style={[styles.text, key === activeKey ? activeTextStyle : inactiveTextStyle]}
              >
                {label}
              </Text>
            </TouchableHighlight>
          ))}
        </View>
      </View>
    </View>
  );
};

export default memo(SwitchNavBar);

export const SwitchNavBarAlt = memo((props: Props) => (
  <SwitchNavBar
    containerStyle={styles.altContainer}
    activePillStyle={styles.altActivePill}
    activeTextStyle={styles.altActiveText}
    inactiveTextStyle={styles.altInactiveText}
    {...props}
  />
));
