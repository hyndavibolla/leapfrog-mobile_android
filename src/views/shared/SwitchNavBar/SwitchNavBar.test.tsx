import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import SwitchNavBar, { Props, SwitchNavBarAlt } from './SwitchNavBar';

describe('SwitchNavBar', () => {
  let props: Props;
  beforeEach(() => {
    props = {
      onPress: jest.fn(),
      activeKey: '1',
      optionList: [
        { key: '1', label: 'one' },
        { key: '2', label: 'two' }
      ]
    };
  });

  it('should render', () => {
    const { toJSON } = render(<SwitchNavBar {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with alt styles', () => {
    const { toJSON } = render(<SwitchNavBarAlt {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with alt styles invalid', () => {
    props.isInvalid = true;
    const { toJSON } = render(<SwitchNavBarAlt {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render without a selected key', () => {
    props.activeKey = null;
    const { toJSON } = render(<SwitchNavBar {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should be pressable', () => {
    const { getByTestId } = render(<SwitchNavBar {...props} />);
    fireEvent.press(getByTestId('switch-nav-bar-btn-2'));
    expect(props.onPress).toBeCalledWith('2');
  });
});
