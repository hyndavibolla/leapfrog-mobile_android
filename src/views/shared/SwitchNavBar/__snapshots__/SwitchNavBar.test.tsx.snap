// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchNavBar should render 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyItems": "center",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyItems": "center",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#EBEDF2",
            "borderBottomLeftRadius": 50,
            "borderBottomRightRadius": 50,
            "borderTopLeftRadius": 50,
            "borderTopRightRadius": 50,
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "space-between",
          },
          undefined,
          Object {
            "padding": "1%",
          },
        ]
      }
    >
      <div
        duration={100}
        style={
          Array [
            Object {
              "borderBottomLeftRadius": 21,
              "borderBottomRightRadius": 21,
              "borderTopLeftRadius": 21,
              "borderTopRightRadius": 21,
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "height": "100%",
              "position": "absolute",
            },
            undefined,
            Object {
              "left": "2%",
              "width": "48%",
            },
            false,
          ]
        }
        transition="left"
      />
      <View
        accessibilityLabel="switch-nav-bar-btn-1"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-1"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                undefined,
              ],
            ]
          }
        >
          one
        </Text>
      </View>
      <View
        accessibilityLabel="switch-nav-bar-btn-2"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-2"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                undefined,
              ],
            ]
          }
        >
          two
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`SwitchNavBar should render with alt styles 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyItems": "center",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyItems": "center",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#EBEDF2",
            "borderBottomLeftRadius": 50,
            "borderBottomRightRadius": 50,
            "borderTopLeftRadius": 50,
            "borderTopRightRadius": 50,
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "space-between",
          },
          Object {
            "backgroundColor": "#FFFFFF",
          },
          Object {
            "padding": "1%",
          },
        ]
      }
    >
      <div
        duration={100}
        style={
          Array [
            Object {
              "borderBottomLeftRadius": 21,
              "borderBottomRightRadius": 21,
              "borderTopLeftRadius": 21,
              "borderTopRightRadius": 21,
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "height": "100%",
              "position": "absolute",
            },
            Object {
              "backgroundColor": "#0066CC",
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
            },
            Object {
              "left": "2%",
              "width": "48%",
            },
            false,
          ]
        }
        transition="left"
      />
      <View
        accessibilityLabel="switch-nav-bar-btn-1"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-1"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#FFFFFF",
                },
              ],
            ]
          }
        >
          one
        </Text>
      </View>
      <View
        accessibilityLabel="switch-nav-bar-btn-2"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-2"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#909BA0",
                },
              ],
            ]
          }
        >
          two
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`SwitchNavBar should render with alt styles invalid 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyItems": "center",
      },
      Object {
        "borderBottomLeftRadius": 4,
        "borderBottomRightRadius": 4,
        "borderColor": "#E02020",
        "borderStyle": "solid",
        "borderTopLeftRadius": 4,
        "borderTopRightRadius": 4,
        "borderWidth": 1,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyItems": "center",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#EBEDF2",
            "borderBottomLeftRadius": 50,
            "borderBottomRightRadius": 50,
            "borderTopLeftRadius": 50,
            "borderTopRightRadius": 50,
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "space-between",
          },
          Object {
            "backgroundColor": "#FFFFFF",
          },
          Object {
            "padding": "1%",
          },
        ]
      }
    >
      <div
        duration={100}
        style={
          Array [
            Object {
              "borderBottomLeftRadius": 21,
              "borderBottomRightRadius": 21,
              "borderTopLeftRadius": 21,
              "borderTopRightRadius": 21,
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "height": "100%",
              "position": "absolute",
            },
            Object {
              "backgroundColor": "#0066CC",
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
            },
            Object {
              "left": "2%",
              "width": "48%",
            },
            false,
          ]
        }
        transition="left"
      />
      <View
        accessibilityLabel="switch-nav-bar-btn-1"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-1"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#FFFFFF",
                },
              ],
            ]
          }
        >
          one
        </Text>
      </View>
      <View
        accessibilityLabel="switch-nav-bar-btn-2"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-2"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#909BA0",
                },
              ],
            ]
          }
        >
          two
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`SwitchNavBar should render without a selected key 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyItems": "center",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyItems": "center",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#EBEDF2",
            "borderBottomLeftRadius": 50,
            "borderBottomRightRadius": 50,
            "borderTopLeftRadius": 50,
            "borderTopRightRadius": 50,
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "space-between",
          },
          undefined,
          Object {
            "padding": "1%",
          },
        ]
      }
    >
      <div
        duration={100}
        style={
          Array [
            Object {
              "borderBottomLeftRadius": 21,
              "borderBottomRightRadius": 21,
              "borderTopLeftRadius": 21,
              "borderTopRightRadius": 21,
            },
            Object {
              "backgroundColor": "#FFFFFF",
              "height": "100%",
              "position": "absolute",
            },
            undefined,
            Object {
              "left": "2%",
              "width": "48%",
            },
            Object {
              "height": 0,
              "width": 0,
            },
          ]
        }
        transition="left"
      />
      <View
        accessibilityLabel="switch-nav-bar-btn-1"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-1"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                undefined,
              ],
            ]
          }
        >
          one
        </Text>
      </View>
      <View
        accessibilityLabel="switch-nav-bar-btn-2"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 12,
            "paddingLeft": 28,
            "paddingRight": 28,
            "paddingTop": 12,
          }
        }
        testID="switch-nav-bar-btn-2"
      >
        <Text
          ellipsizeMode="tail"
          font="SFProDisplay-Heavy"
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Heavy",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#262C2F",
                  "fontSize": 12,
                  "letterSpacing": 1,
                  "lineHeight": 14,
                  "textTransform": "uppercase",
                },
                undefined,
              ],
            ]
          }
        >
          two
        </Text>
      </View>
    </View>
  </View>
</View>
`;
